go-controller/pkg/ovn/controller/apbroute/external_controller_policy.go

```diff
@@ -6,39 +6,266 @@ import (
 	"reflect"
 	"strings"
 
-	v1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	ktypes "k8s.io/apimachinery/pkg/types"
 	kerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
 	"k8s.io/klog/v2"
 
 	adminpolicybasedrouteapi "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/adminpolicybasedroute/v1"
+	adminpolicybasedroutelisters "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/adminpolicybasedroute/v1/apis/listers/adminpolicybasedroute/v1"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
 )
 
-// processAddPolicy takes in a new policy and applies it. To do that, it aggregates the IPs from the static hops and retrieves the IPs from the pods resulting from applying the
+func (m *externalPolicyManager) syncRoutePolicy(routePolicy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute,
+	routeLister adminpolicybasedroutelisters.AdminPolicyBasedExternalRouteLister, routeQueue workqueue.RateLimitingInterface) error {
+	_, err := routeLister.Get(routePolicy.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	if apierrors.IsNotFound(err) || !routePolicy.DeletionTimestamp.IsZero() {
+		// DELETE use case
+		klog.Infof("Deleting policy %s", routePolicy.Name)
+		err = m.processDeletePolicy(routePolicy.Name, routeQueue)
+		if err != nil {
+			return fmt.Errorf("failed to delete Admin Policy Based External Route %s:%w", routePolicy.Name, err)
+		}
+		return nil
+	}
+	currentPolicy, found, markedForDeletion := m.getRoutePolicyFromCache(routePolicy.Name)
```

Comment:
I also checked the repair code, and I see that it doesn't return an error, so the initial cache build may fail, and it looks like that can cause some problems, e.g. if the policy existed before startup, but it wasn't added by repair logic to the cache, `processAddPolicy` will be called instead of `processUpdatePolicy`. I am missing something or is it the case?

Reply:
will examine the cache issues this afternoon

Reply:
Alright so @jordigilh and I walked through the code this afternoon. This is what we identified and what will be fixed:

1. As you mentioned @npinaeva the initialization of the cache is done in the repair. This has a bug because there is no guarantee that the policies initialized in the cache have actually been configured in NBDB. Therefore we could accidentally shortcut with the cache on ADD/UPDATE event and the policy is never configured in NBDB. - @jordigilh will file a new PR that removes initializing the cache in repair, and just let the add events from the informer build the cache

2. With your example of spec 1->spec2->spec1. There are a couple of things here:
2a. Now with single event loop based on keys, the latest policy is fetched when processed from informer cache, and compared to the version that is in the internal policy cache. The policy cache is not updated until the end of the processing so we should be safe here.
2b. However, the namespace cache needs to only be updated at the end of policy processing.
2c. Also, there is the issue of halfway configured NBDB objects when a failure occurs. @jordigilh will change this to bundle ops together in a single txn.
2d. Therefore we will have: lock namespace cache, execute 1 NBDB txn, if err ==nil update namespace and policy cache . @jordigilh will file a separate PR to fix this problem

3. There is another problem you identified where we are missing queuing keys for namespace events where the dynamic gw selector might be a namespace.  @jordigilh will fix that in this PR

4. Finally, there is an issue with how pods are handled. The current code relies on the legacy code in addLogicalPort and delLogicalPort to handle adding routes for served pods as they are created and deleted. However this has the potential to race with APB controller trying to apply policy for these pods at the same time. To conform to the design of a single loop handling everything we will:
4a. Change the legacy path to only configure NBDB for non-overlapping IPs that are in the annotation.
4b. Add code to APB to handle configuring served pods, as well as any overlapping IPs that also exist in annotations. This will be done in a separate PR.